### ISSUE TITLE
Bizarre React import type issue (works when * but not when specific types are imported)

### ISSUE THREAD

**Issue by EvHaus** [facebook/flow#4841]
I've encountered a really bizarre bug with Flow 0.54.0. It seems that the way you import React types is important.

The following code works just fine (https://flow.org/try/#0PTAEAEDMBsHsHcBQiCWBbADrATgF1AFSgCGAzqAEoCmxAxvpNrGqAOTY32sDcyuAnhiqgACkwykAkgDsMAV1wAVQcIC8oAN6IAkMQBcoaXLQAjKtkQBfXogFDR40gGFY06VXpUAJsvvqt2iYGRqbmVjZ2wmKwEi6mKO4+KqDq0RIy8krJAGQOMc6u7p5JQja0rqT41ACOcigcpACCbrC4xLgoFQASAPJOKaCQctL0ndKgADxppAbTcSYJ3r5UAHygABQ6cVik3tuuVNK4BtR0uAB0++5Hy1OO84slwrnTK4gAlCecF1eHWUJ3fJrAIcXBybDjWjQMjkKgAD1why85FO9HOInBVF+R0BEmBOm0HGkXnMG0+JGk-FAmgJhKoYIhk32uy82PwGnOnNwAAsUKRzhhHJZQMAVrxtNpLDprOFkOVpJVKFRavUqE0Wm0OhUBkMRlrxriZnl0rIFMs1pttMy9swsNdjkqzpdbQcbipDQ9EstQC9HG9yaifi77bdXjTtPLFQB1bDEDBCVnBv4DGp1BrNaStdpjUi9Jzra2JzCu3DvcWg8GQ6GkWEIpEo77ozFsw34iVEknYMkGYiU6kBdv0yuTGNxhNszSc848vkCoWgEyqDQAJmFovFkulvGsQA):

```javascript
// @flow

import * as React from 'react';

type PropsInputType = {
	a: number
};

type PropsConnectedType = {
	b: number
};

type PropsCombinedType = PropsInputType & PropsConnectedType;

const RequiresAnnotationsHOC = function <Props: PropsCombinedType> (
	ComposedComponent: React.ComponentType<PropsCombinedType & Props>
): React.ComponentType<Props> {
	return class extends React.PureComponent<Props> {
		render (): any  {
			return <ComposedComponent {...this.props} />;
		}
	};
};

const RequiresAnnotations = function <Props: PropsInputType> (
	ComposedComponent: React.ComponentType<PropsCombinedType & Props>
): React.ComponentType<Props> {
	const WrappedComponent = RequiresAnnotationsHOC(ComposedComponent);
	return class extends React.PureComponent<Props> {
		render (): any  {
			return <WrappedComponent {...this.props} b={2} />;
		}
	};
};
```

However, this code (https://flow.org/try/#0PTAEAEDMBsHsHcBQiCWBbADrATgF1AEoCmAhgMa4A0oA3rgJ4ZGgDCsmsAdkZ7gCqMi1Bk1ABRaETQ8qoAAoBXbETYduvAL6hI2dqADky8rn0BuZCOZzdGAM4BJThgX9BoALy1EASBIAuUE4FNAAjImxEDXNES3kbWzZObgoiABMBUU8aHxCAoNDwyOjY61g7VRCUbnS3T1K7R2dXUQAyOLKErmTcNIyiaLIuW3xiAEcFFGVbAEEk2FwSXBQhgAkAeRYPbQVOCmXOUAAeetsAk4qq3sEAPlAACh9VLFs0p64ZALf1ZqJj+Ivqn1QG0TtdEABKT7sLDfPp-Dq3bLeZS4JQHMjQEi2WygIgADx6nFSOMUyi+MnhdkRPmRPFS4XukNAJE49FAXm8nJRaKObxeqXJvFoADpRbgABYoWzCjDxLTAa7mTkaHxRIrIQacYaEIjjSZEGZzBZLIZbSA7PZcI4nM7xRouPq3B7ePmvaHvXhQtQyOHndiVQFuEHxMFMwU-Sm2aneTXagDq2BIGCYAvd3y2YwmU1mnHmi32tnWLDurtT3t44KV3Ow6Mx2NxBLpJKUKjTFNBHNpRIZdyZLLZndpqJrRwTSZT4ZFYsl0tlHS0IXcNAATPLFTSVd41VEgA):

```javascript
// @flow

import React, {type ComponentType, type Element, PureComponent} from 'react';

type PropsInputType = {
	a: number
};

type PropsConnectedType = {
	b: number
};

type PropsCombinedType = PropsInputType & PropsConnectedType;

const RequiresAnnotationsHOC = function <Props: PropsCombinedType> (
	ComposedComponent: ComponentType<PropsCombinedType & Props>
): ComponentType<Props> {
	return class extends PureComponent<Props> {
		render (): any  {
			return <ComposedComponent {...this.props} />;
		}
	};
};

const RequiresAnnotations = function <Props: PropsInputType> (
	ComposedComponent: ComponentType<PropsCombinedType & Props>
): ComponentType<Props> {
	const WrappedComponent = RequiresAnnotationsHOC(ComposedComponent);
	return class extends PureComponent<Props> {
		render (): any  {
			return <WrappedComponent {...this.props} b={2} />;
		}
	};
};
```

Will fail with:

```
13: type PropsCombinedType = PropsInputType & PropsConnectedType;
                                              ^ property `b`. Property not found in
20: 			return <ComposedComponent {...this.props} />;
              ^ props of React element `ComposedComponent`
```

If you look closely the ONLY difference is how the types are imported. In the top example, they are imported as `import * as React from 'react';` whereas below they are imported as `import React, {type ComponentType, type Element, PureComponent} from 'react';`.

This is still an issue with flow 0.56.0, and was introduced in flow 0.54.0

Any ideas?

**Comment by asolove** [comment]
Huh, also rearranging them also fixes it:

```js
import React, {PureComponent, type ComponentType, type Element} from 'react';
```

Not sure why though.

**Comment by tdeekens** [comment]
This literally ended a hour long journey @asolove. In combination with rollup importing `import * as React from 'react'` instead of `import React from 'react'` causes an UMD bundle get receive code in `"var React__default = React['default'];"` rather than `"var React__default = 'default' in React ? React['default'] : React;"` in the factory of the UMD bundle. Causing the my library to not work. Being aware that I could import types separately from the packages which is not mentioned in the docs of flow solved this for me. Thanks a lot!

**Comment by EvHaus** [comment]
Closing. This is no longer occurring with Flow 0.91.0. I think it was fixed around 0.89.0 or so.
